helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -1261,31 +1257,75 @@ private void processDataOrChildChange(WatchedEvent event, long notificationTime)
 
     if (event.getType() == EventType.NodeDataChanged || event.getType() == EventType.NodeDeleted
         || event.getType() == EventType.NodeCreated) {
+      boolean pathExists = event.getType() != EventType.NodeDeleted;
```

Comment:
This might be a potential bug. For example,
T1: node deleted, and NodeDeleted is sent.
T2: node is re-created
T3: The NodeDeleted event is processed, pathExists is set to false. And it is **NOT** the truth!

This might be OK, since we will have another event coming later. But it may introduce a substantial change of the ZkClient behavior. So I would prefer if we don't do it. The minimum fix itself does not require this enhancement. Please correct me if I am wrong.

Reply:
It's an interesting case though I don't think it will introduce any new substantial change to the behavior. Given ZK's consistency, the node recreation event will guarantee to arrive later than the node deletion event. Let me explain fundamentally what happens on `handleNodeDeleted`, it unsubscribes the listener entry in zkClient (it can be confirmed that all operations are all in-memory operations without any ZK relations); 
When the T2 node re-creation event gets processed, it will re-subscribe the listener just like it has never seen it before.

Reply:
I think you don't have to check pathExits here and pass it to fireDataChangedEvents(). Instead, you can check the condition within `fireDataChangedEvents`. The pro is the api can avoid the boolean param and 2 logic branches in `fireDataChangedEvents`. If it is a DELETED event, you will call `handleDataDeleted`.

Reply:
No longer valid comment since the new change doesn't modify the existing codes that much